agent/agent_endpoint.go

```diff
@@ -359,7 +360,11 @@ func (s *HTTPHandlers) AgentServices(resp http.ResponseWriter, req *http.Request
 // blocking watch using hash-based blocking.
 func (s *HTTPHandlers) AgentService(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
 	// Get the proxy ID. Note that this is the ID of a proxy's service instance.
-	id := strings.TrimPrefix(req.URL.Path, "/v1/agent/service/")
+	id, err := url.PathUnescape(strings.TrimPrefix(req.URL.Path, "/v1/agent/service/"))
+	if err != nil {
+		s.agent.logger.Error(fmt.Sprintf("failure in unescaping path param: %s", err.Error()))
+		id = strings.TrimPrefix(req.URL.Path, "/v1/agent/service/")
+	}
```

Comment:
I recommend that we simplify this to make it easier to maintain and apply consistently to all endpoints (remove redundant operations and boilerplate). I think this can be solved by abstracting into a function call like:
```
func GetPathSuffixUnescaped(path string, prefixToTrim string) (string, error) {
  // The suffix may be URL-encoded, so attempt to decode
  suffixRaw := strings.TrimPrefix(path, prefixToTrim)
  suffixUnescaped, err := url.PathUnescape(suffixRaw)
  
  if err != nil {
    return suffixRaw, fmt.Errorf("failure in unescaping path param %q: %v", suffixRaw, err)
  }
  
  return suffixUnescaped, nil
}
```

Then, each endpoint replacement would look something like this:
```suggestion
	id, err := GetPathSuffixUnescaped(req.URL.Path, "/v1/agent/service/")
	if err != nil {
		s.agent.logger.Error(err)
	}
```

---

@kisunji : I'm not sure the best place in the source code to put a shared function like this. Thoughts on that? And whether you agree with this code change suggestion?

@littlestar642: are you interested in making such an enhancement to the PR? If not (we realize some time has passed since you posted this), let us know!

Reply:
Great suggestion! I think `agent/http.go` is a reasonable place for a general function that can be reused by all endpoints.

Reply:
Thanks - change made! I also added a unit test for that function.